### ISSUE TITLE
parallelize nomos and copyright agent when scanning multiple files on multicore machines

### ISSUE THREAD

**Issue by yaobinshi** [fossology/fossology#208]
---

Author Name: **armijn hemel** 
Original Redmine Issue: 3909, http://www.fossology.org/issues/3909
Original Date: 2013/01/15
Original Assignee: larry shi

---

when scanning multiple files at once with nomos or the copyright agent it might make sense to partially parallellize the scanning. In the scripts to generate the Binary Analysis Tool database I have very good experiences with this: I have a list of files, which I divide in chunks (chunks of 1 for the copyright agent, chunks of 10 for nomos since it can process multiple files at once), then launch a number of processes for the chunks.

Since scanning a file is independent from scanning another file this would be a perfect candidate for parallel programming.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **armijn hemel** 
Original Date: 2013/06/06 08:20:00 +0000

---

Of course, it would make sense to make this configurable, as to prevent too many processes from being launched at the same time without the user knowing.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **larry shi** 
Original Date: 2013/06/07 02:46:01 +0000

---

parallel programming is interesting.

right now, some agents, e.g. nomos copyright can run in parellel, you can test them with a large package upload.

for a single agent running in parallel, right now, only ununpack agent has this functionality. 
if nomos/copyright want to have this functionality, that will be great, but will need lots of logic change.


also for me, single agent running in parallel has a lower priority.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **Bob Gobeille** 
Original Date: 2013/08/02 20:41:06 +0000

---

Jobs (uploads) can run in parallel but nomos runs single threaded on a single upload.  FWIW, before the 2.0 scheduler, agents working on single uploads were run parallel.  However, in that implementation there was a lot of traffic between the scheduler and agent(s) that limited its usefulness.

This is still a good idea but needs some analysis to figure out the best way to do it.  It won't get done for 2.3 (insufficient time available).

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **Bob Gobeille** 
Original Date: 2014/02/03 17:15:25 +0000

---

We have had requests to improve the throughput of nomos scans from both Silvano and Bryan S.  So I'm raising the priority.  It is still marked for 2.5 but I don't expect it to receive any time until 2.6

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **Bob Gobeille** 
Original Date: 2014/06/25 22:41:34 +0000

---

Also see #7137

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **Bob Gobeille** 
Original Date: 2014/08/25 16:19:41 +0000

---

Need to optimize performance.
Change dictionary initialization to after evaluating switches.  Don't init dictionaries that are not needed.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **armijn hemel** 
Original Date: 2014/08/25 18:51:06 +0000

---

Bob Gobeille wrote:

> Change dictionary initialization to after evaluating switches.  Don't init dictionaries that are not needed.


To expand a bit on this: the dictionaries are not used for e-mail and URL scanning. If someone is only interested in any of these two and not copyright extraction, then initialization of the dicts is not necessary and can be skipped.

**Comment by mcjaeger** [comment]
it is an oldie but goldie especially for those who use nomos (or other fossology agents)  from the commandline

**Comment by mcjaeger** [comment]
it is apparently no one is catching up for this issue for a long time.
